design-proposals/klip-7.md

```diff
@@ -0,0 +1,176 @@
+
+The Execution Plan is a DEG of Execution Steps. Each Execution Step represents a call to Kafka Streams, along with context about what the resulting Streams operator is doing. We can build the Execution Plan at the SchemaKStream layer.
+
+For example, the call to SchemaKStream.select would look something like:
+
+```
+public SchemaKStream<K> select(
+    final List<SelectExpression> selectExpressions,
+    final QueryContext.Stacker contextStacker,
+    final ProcessingLogContext processingLogContext) {
+  final ExecutionStep step = new StreamSelect(this.executionStep, selectExpressions);
+  return new SchemaKStream<>(
+      step.build(),
+      ksqlConfig,
+      functionRegistry,
+      contextStacker.getQueryContext(),
+      step
+  );
+}
+```
+
+Where StreamSelect looks like:
+
+```
+public class StreamSelect implements ExecutionStep {
+  ...
+  private final ExecutionStep source;
+  private final List<SelectExpression> selectExpressions;
+
+  @JsonCreator
+  public StreamSelect(
+      ...
+      @JsonProperty("source") final ExecutionStep source,
+      @JsonProperty("selectExpressions") final List<SelectExpression> selectExpressions) {
+    this.properties = Objects.requireNonNull(properties, "properties");
+    this.source = Objects.requireNonNull(source, "source");
+    this.selectExpressions = ImmutableList.copyOf(selectExpressions);
+  }
+
+  public List<SelectExpression> getSelectExpressions() {
+    return selectExpressions;
+  }
+
+  @Override
+  public List<ExecutionStep> getSources() {
+    return Collections.singletonList(source);
+  }
+
+  @Override
+  public KStream build(final KStream input) {
+      final Selection selection = new Selection(selectExpressions);
+      return kstream.mapValues(selection.getSelectValueMapper()),
+  }
+}
+```
+
+I've written up a basic implementation of Execution Plans which includes all the Execution Steps we'd need, and can be found here (https://github.com/confluentinc/ksql/blob/ksql-execution-plans/ksql-common/src/main/java/io/confluent/ksql/util/KsqlConfig.java) (steps are under ksql-engine/src/main/java/io/confluent/ksql/structured/execution):
+
+#### Serializability
+The Execution Plan needs to be JSON-serializable. This way, we can persist the plan generated for each case in our Query Translation tests, and verify that it is kept the same across versions.
+
+#### Versioning
+We may need to version Execution Steps. Remember that a step is meant to represent a basic stream processing operation performed by a call to Kafka Streams. We may want to evolve how we perform a given operation in a way that is incompatible across versions. If a developer needs to modify a step in an incompatible way, they can do so by writing a new step type and advancing the version.
```

Comment:
Versioning would take place per Execution Step, right? What would that look like? Would each version have to be a new file? Trying to understand how this sort of version management typically works, mostly out of curiosity.

Reply:
> Versioning would take place per Execution Step, right?

Yes, that's correct.

> What would that look like?

It would just be a naming scheme for the classes. So let's say we wanted to make an incompatible change to how we do a projection, the new execution step for doing a projection could be called `StreamSelectV2`